### PR TITLE
Ensure that the compilers produce double.NaN values in IEEE canonical form.

### LINKED ISSUE #30439: Compilers should normalize NaNs produced for 0.0/0.0 for determinism
Apparently some CLRs have issues such that `0.0 / 0.0` produces different results (specifically, NaNs with different bit patterns) on different platforms.  To work around this, Roslyn compilers should normalize NaNs so that we produce the same result on all platforms.

See also https://github.com/dotnet/corefx/pull/32741 which adds tests for possibly problematic runtimes.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #30439

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs [hunk 1]
```diff
@@ -16692,5 +16692,36 @@ static void Main()
...
+        const double inf = 1.0 / 0.0;
+        CheckNaN(inf + double.NaN);
+        CheckNaN(inf - double.NaN);
+        CheckNaN(-double.NaN);
+    }
+
+    static void CheckNaN(double nan)
+    {
+        const long expected = unchecked((long)0xFFF8000000000000UL);
+        long actual = BitConverter.DoubleToInt64Bits(double.NaN);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: Why depend on a helper here vs. just using unsafe?

Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: This should be `nan` not `double.NaN`

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb [hunk 2]
```diff
@@ -14753,6 +14753,36 @@ End Module
...
+        CheckNaN(0.0 / -0.0)
+        Dim inf As Double = 1.0 / 0.0
+        CheckNaN(inf + Double.NaN)
+        CheckNaN(inf - Double.NaN)
+        CheckNaN(-Double.NaN)
+    End Sub
+
+    Shared Sub CheckNaN(nan As Double)
+        Dim expected As Long = &amp;HFFF8000000000000
+        Dim actual As Long = System.BitConverter.DoubleToInt64Bits(Double.NaN)
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb]: This should be `nan` as well.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs [hunk 3]
```diff
@@ -16692,5 +16692,36 @@ static void Main()
...
+            string source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        CheckNaN(double.NaN);
+        CheckNaN(float.NaN);
+        CheckNaN(0.0 / 0.0);
+        CheckNaN(0.0 / -0.0);
```
Comment (tannergooding) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: Is testing `-0.0 / 0.0` and `-0.0 / -0.0` also important?

File: src/Compilers/Core/Portable/ConstantValueSpecialized.cs [hunk 4]
```diff
@@ -8,6 +8,19 @@ namespace Microsoft.CodeAnalysis
 {
     internal partial class ConstantValue
     {
+        /// <summary>
```
Comment (tannergooding) [inline src/Compilers/Core/Portable/ConstantValueSpecialized.cs]: It is probably worth noting that we choose the "cannonical" NaN with the sign bit set. Both positive and negative NaN, with no diagnostic payload are considered "cannonical".

File: src/Compilers/Core/Portable/ConstantValueSpecialized.cs [hunk 5]
```diff
@@ -753,6 +771,11 @@ private sealed class ConstantValueSingle : ConstantValueDiscriminated
             public ConstantValueSingle(double value)
                 : base(ConstantValueTypeDiscriminator.Single)
             {
+                if (double.IsNaN(value))
+                {
+                    value = _s_IEEE_canonical_NaN;
```
Comment (tannergooding) [inline src/Compilers/Core/Portable/ConstantValueSpecialized.cs]: This is relying on the underlying hardware to exactly preserve both the sign and the "payload" bits when converting from double to single, which is not required by the spec.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: This helper reads the bits underlying the double. Using unsafe I'd have to define a struct that overlays a double and a long, or make a pointer to a local variable and cast the pointer. But that might not be correct depending on the endianness of the host running the compiler; I'd have to write compensating code that is resilient to host endianness. This is just a simple and clear way to do it.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: There are a huge number of ways of producing NaN values. I'm just trying a small representative sample of them.

File: src/Compilers/Core/Portable/ConstantValueSpecialized.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/Core/Portable/ConstantValueSpecialized.cs]: It is required by the spec that a float canonical NaN is required to round trip (without loss) to and from a double NaN [IEEE Std 754-2008 section 6.2.3].

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs [hunk 6]
```diff
@@ -16692,5 +16692,36 @@ static void Main()
...
+        public void NormalizedNaN()
+        {
+            string source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        CheckNaN(double.NaN);
+        CheckNaN(float.NaN);
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: I will add tests for float, but as I pointed out the requirements of section 6.2.3 of IEEE 754 do indeed provide the needed guarantees for canonical NaN values.

File: src/Compilers/Core/Portable/ConstantValueSpecialized.cs (hunk 5 above)
Comment (tannergooding) [reply src/Compilers/Core/Portable/ConstantValueSpecialized.cs]: Right but 6.3 explicitly specifies that for both inputs and results, the standard does not interpret the sign of the NaN. And 10.4 specifies that changing the sign of a NaN is allowed as a "value-changing transformation which preserves the literal meaning of the source code".

**Comment by gafter** [comment]
I believe I have addressed all comments. @tannergooding @jaredpar Do you have any further comments?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs [hunk 7]
```diff
@@ -16692,5 +16692,53 @@ static void Main()
...
+        const long expected = unchecked((long)0xFFF8000000000000UL);
+        long actual = BitConverter.DoubleToInt64Bits(nan);
+        if (expected != actual)
+            throw new Exception($""expected=0X{expected:X} actual=0X{actual:X}"");
+    }
+
+    static unsafe void CheckNaN(float nan)
+    {
+        const int expected = unchecked((int)0xFFC00000U);
+        void* p = &nan;
```
Comment (tannergooding) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: Just wondering, why this rather than just `int actual = *(int*)&nan`?

Comment (gafter) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: No particular reason.

**Comment by gafter** [comment]
test ubuntu_16_debug_prtest please

**Comment by gafter** [comment]
test ubuntu_16_debug_prtest please

**Comment by gafter** [comment]
PR testing appears to be broken.

**Comment by gafter** [comment]
Looks like something in the testing infrastructure is broken. `ubuntu_16_debug_prtest` fails because

> 14:49:53 Shutting down Razor build server (process 58421)...
> 14:49:53 Razor build server (process 58421) failed to shut down: The shutdown command failed: The application to execute does not exist: '/mnt/j/workspace/dotnet_cli/release_2.2.1xx/release_linux_x64_prtest/.nuget/packages/microsoft.aspnetcore.razor.design/2.2.0-preview3-35497/tools/rzc.dll'

I'm going to retarget master since it is getting late for preview 1.

**Comment by jaredpar** [comment]
Everything is green now. Are we good to merge?

**Pull request merged**